### DISCUSSION TITLE
Extracting connectivity graph from FLAX or JAX

### DISCUSSION THREAD

**Discussion by mattiasmar** [google/flax#1718] [Q&A] [answered]
Hi,
Looking for a way to extract the computational graph from Flax (or JAX). More precisely I want to gather the connectivity graph and the dimensions of the tensors communicated between operators. Any ideas / pointers?

**Comment by jheek** [comment, marked as answer]
I think what you are looking for is `jax.make_jaxpr` this gives you all the jax operations. It has a list of operations `eqns` each equation has invars/outvars and a var has a `.aval` field which will typically have a shape so you can do things like:
```
jaxpr = jax.make_jaxpr(fn)(*inputs)
inputs = jaxpr.eqns[i].invars
print([x.aval.shape for x in inputs])
```

**Reply by mattiasmar** [reply]
@jheek Thanks.   
Do you know how one could align that output with that of the pytree given by: `_, treedef = jax.tree_flatten(state.params)` ?
It is the graph behind the operators and tensors of this `treedef `that I'm out after.

**Reply by jheek** [reply]
the invars of the jaxpr will align with jax.tree_flatten(fn_args) So you could just call treedef.unflatten(jaxpr.invars) and retrieve the invars related to params from that

**Reply by mattiasmar** [reply]
I took the example from [/HuggingFace/vision/run_image_classification.py](https://github.com/huggingface/transformers/tree/master/examples/flax/vision)
On line 485 (where it says: `state, train_metric = p_train_step(state, batch)`) I call `jaxpr = jax.make_jaxpr(p_train_step)(state, batch)`  and get that `jaxpr.eqns` is a list of length 1 and that jaxpr.eqns[0].invars is of length 606.
At the same time the treedef of state.params has only 200 leaves. Thus treedef.unflatten(jaxpr.eqns[0].invars) fails because of the dimensions mismatch `(ValueError: Too many leaves for PyTreeDef; expected 200.)`

```
jaxpr = jax.make_jaxpr(p_train_step)(state, batch)
_, treedef = jax.tree_flatten(state.params)  #treedef.num_leaves  = 200
jaxpr = jax.make_jaxpr(p_train_step)(state, batch) # len( jaxpr.eqns[0].invars) = 606
treedef.unflatten(jaxpr.eqns[0].invars)
```

**Reply by jheek** [reply]
The treedef in this case is `treedef = jax.tree_structure((state, batch))` and then you use `param_invars = treedef.unflatten(jaxpr.jaxpr.invars)[0].params` (note the .jaxpr)

**Reply by mattiasmar** [reply]
Thanks again @jheek. I see now how I can get the dimensions of input and output tensors of the operators listed in the treedef ( as returned by `jax.tree_structure((state, batch)) )`.
The piece missing for me now is the connectivity.  From what I can tell from running the above code snippets is that param_invars does not match param_outvars (i.e. one doesn't contain the tensor names of other).  Do you know what in JAX could give me the edges between operators?

**Reply by jheek** [reply]
The name var doesn't exactly mean what you expect I think. JAXPR vars aren't actually updated and they are more like identifiers. Each output of an operation is written to a unique var. To find the connective you have to check the invars/outvars of the equations. This is further complicated by the fact that some equations can contain jaxprs inside of them (calls,  scans, etc.)

**Reply by avital** [reply]
@jheek -- should we also consider a more "Flax native" graph description, sort of like this in Haiku: https://dm-haiku.readthedocs.io/en/latest/notebooks/visualization.html?

**Reply by jheek** [reply]
Actually the visualization in Haiku is turning a JAXPR -> dot graph. So that could also just be part of JAX. The table is somewhat similair to the `__repr__` impl in Flax although a table is definitely nice. But if I understand correctly a table of layers is not what @mattiasmar is looking for

**Reply by wztdream** [reply]
Is it possible to visualize the computation graph using tensorboard, as both jax/flax and tensorflow are from google, I suppose it will be much easier to solve the incompatible part.

**Reply by jheek** [reply]
Tensorboard doesn't support a XLA computation graph it only shows a tf graph. The haiku visualization module should work on any JAX function. And I think it will also support the Flax metadata (so it will groups ops belonging to the same flax Module).

**Reply by mattiasmar** [reply]
@jheek  the haiku [to_dot](https://github.com/deepmind/dm-haiku/blob/main/haiku/_src/dot.py) source code is dependent on the haiku [Module class]([/dm-haiku/blob/main/haiku/_src/module.py](https://github.com/deepmind/dm-haiku/blob/main/haiku/_src/module.py)). What would be Flax' replacement for it?

**Reply by jheek** [reply]
Is it? I think you can just call `to_dot(flax_module.apply)(variables, inputs)`. I haven't tried it though

**Reply by mattiasmar** [reply]
I tested it in [HuggingFace\vision\run_image_classification.py](https://github.com/huggingface/transformers/tree/master/examples/flax/vision) and got the below error which appears to originate from the [linen.module line 1108](https://github.com/google/flax/blob/4acd0fe7b2b44d6a38e226faef5006aa1c45d6bd/flax/linen/module.py#L1108) where the first argument is a hardcoded empty dict.

```
from haiku._src import dot
dot.to_dot(model.module.apply)(state,batch)
```

```
   File "/usr/local/lib/python3.8/dist-packages/flax/linen/module.py", line 1074, in apply
    return apply(
  File "/usr/local/lib/python3.8/dist-packages/flax/core/scope.py", line 729, in wrapper
    with bind(variables, rngs=rngs, mutable=mutable).temporary() as root:
  File "/usr/local/lib/python3.8/dist-packages/flax/core/scope.py", line 703, in bind
    if not _is_valid_variables(variables):
  File "/usr/local/lib/python3.8/dist-packages/flax/core/scope.py", line 783, in _is_valid_variables
    for name, col in variables.items():
AttributeError: 'TrainState' object has no attribute 'items'
```

**Reply by mattiasmar** [reply]
@jheek  When applying the to_dot method on the module of  the Flask Imagenet Example (in the [collab](https://colab.research.google.com/github/google/flax/blob/main/examples/imagenet/imagenet.ipynb))

```
from haiku._src import dot
dot.to_dot(model.apply)({'params': state.params, 'batch_stats': state.batch_stats}, batch['image'][:128])
```

one gets the error:

```
TypeError                                 Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/haiku/_src/dot.py in aval(self)
    183   @property
    184   def aval(self):
--> 185     return jax.core.get_aval(self.val)
    186 
    187   def full_lower(self):
...
type <class 'tensorflow.python.framework.ops.EagerTensor'> is not a valid JAX type
```

**Reply by jheek** [reply]
Looks like you passed tf tensor instead of a numpy or jax array?

**Reply by mattiasmar** [reply]
Indeed, and if I convert the batch object to a numpy array I get a `ModifyScopeVariableError`.
Do you have  a working example of the use of to_dot on a flax module that you could share?

**Reply by marcvanzee** [reply]
@mattiasmar the Colab you linked doesn't seem to show me your error.
Could you give us a minimal example in a Colab where the `ModifyScopeVariableError` occurs?
That will make it much easier for us to help you.

**Reply by mattiasmar** [reply]
In the original collab,  just before the "Inference" section add:
```
!pip install -q dm-haiku
from haiku._src import dot
batch = next(iter(eval_ds))
dot.to_dot(model.apply)({'params': state.params, 'batch_stats': state.batch_stats}, batch['image'][:128].numpy())
```
...and you should see the error that I mentioned.

**Reply by mattiasmar** [reply]
@marcvanzee I created a reduced copy of the original Colab [here](https://colab.research.google.com/drive/1bV8pS7zYCSLbl59IQWYq-uufnXUQLgj7?usp=sharing) demonstrating the error when calling to_dot.

**Reply by marcvanzee** [reply]
The error states that the variable collection `batch_stats` should be mutable. However, by default only the `params` collection is mutable, so you should add that as a keyword argument to the apply functions. Since `to_dot` wraps `model.apply` inside a function it doesn't seem possible to pass kwargs to it, but the following works:

```py
dot.to_dot(functools.partial(model.apply, mutable=['batch_stats']))(
    {'params': state.params, 'batch_stats': state.batch_stats},
    batch['image'][:128])
```

This outputs something that I suppose you should pass to `graphviz.Source(out)`.

**Reply by mattiasmar** [reply]
Thanks @marcvanzee. Your code change makes a lot of sense. I had a missed to educate myself on that parameter.

I tried the notebook in the TPU Collab but it caused the whole runtime to crash and restart. The logs says only:
`2022-01-04 13:13:25.808480: W tensorflow/core/framework/cpu_allocator_impl.cc:82] Allocation of 308281344 exceeds 10% of free system memory.`

Locally though, on my own GPU system, with the code change I now get a different error than before (which is a good thing! )
```
RuntimeError: UNKNOWN: Failed to determine best cudnn convolution algorithm for:
%cudnn-conv = (f32[32,112,112,64]{2,1,3,0}, u8[0]{0}) custom-call(f32[32,224,224,3]{2,1,3,0} %copy.3, f32[7,7,3,64]{1,0,2,3} %copy.4), window={size=7x7 stride=2x2 pad=3_3x3_3}, dim_labels=b01f_01io->b01f, custom_call_target="__cudnn$convForward", metadata={op_type="conv_general_dilated" op_name="jit(conv_general_dilated)/conv_general_dilated[\n  batch_group_count=1\n  dimension_numbers=ConvDimensionNumbers(lhs_spec=(0, 3, 1, 2), rhs_spec=(3, 2, 0, 1), out_spec=(0, 3, 1, 2))\n  feature_group_count=1\n  lhs_dilation=(1, 1)\n  lhs_shape=(32, 224, 224, 3)\n  padding=((3, 3), (3, 3))\n  precision=None\n  preferred_element_type=None\n  rhs_dilation=(1, 1)\n  rhs_shape=(7, 7, 3, 64)\n  window_strides=(2, 2)\n]" source_file="/usr/local/lib/python3.8/dist-packages/haiku/_src/dot.py" source_line=204}, backend_config="{\"conv_result_scale\":1,\"activation_mode\":\"0\",\"side_input_scale\":0}"

Original error: UNIMPLEMENTED: DNN library is not found.
```

```
tf-docker ~ > echo $LD_LIBRARY_PATH
/usr/local/cuda-11.4/lib64:/usr/local/cuda-11.0/targets/x86_64-linux/lib:/usr/local/cuda/extras/CUPTI/lib64:/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
tf-docker ~ > echo $PATH
/usr/local/cuda-11.4/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

Is your guess that this error is related to JAX, Flax or Haiku? From other git issues with this error, setting the LD_LIBRARY_PATH has been suggested as a solution, but it doesn't seem to have done the trick for me.

**Reply by marcvanzee** [reply]
As the error says, the image is too big to fit in memory. Why don't you start with a simple model, like this?

```py
from flax import linen as nn
from jax import random
import jax.numpy as jnp
from haiku._src import dot
import graphviz

class Foo(nn.Module):
  @nn.compact
  def __call__(self, x):
    for _ in range(3):
      x = nn.Dense(4)(x)
    return x

inputs = jnp.zeros((3, 4))
variables = Foo().init(random.PRNGKey(0), inputs)
graphviz.Source(dot.to_dot(Foo().apply)(variables, inputs))
```

Outputs:
<img width="1106" alt="Screenshot 2022-01-04 at 14 53 02" src="https://user-images.githubusercontent.com/180100/148069253-30e8ab0a-c821-4193-bb22-1ce8aca98fc2.png">

**Reply by mattiasmar** [reply]
Excellent. I got it to work now, also for the ImageNet example. Thanks so much for your help.
One of course actually doesn't actually need the accelerator (TPU/GPU) at all to generate the graph and so it is easy to avoid the OOM error.
I created a  minimalistic [Collab ](https://colab.research.google.com/drive/1bV8pS7zYCSLbl59IQWYq-uufnXUQLgj7?usp=sharing)for the ResNet example. Perhaps that could help someone in the future.
![image](https://user-images.githubusercontent.com/48177580/148087039-c7b22272-fdbe-48a1-a115-a39fd459d14c.png)

**Comment by avital** [comment]
> Is it possible to visualize the computation graph using tensorboard, as
both jax/flax and tensorflow are from google, I suppose it will be much
easier to solve the incompatible part.

Actually, I believe I recall someone using jax2tf (which includes Flax
metadata because of Flax's use of named_call, which become tf.name_scopes)
and loading a graph into the TensorBoard visualization that looked pretty
good. If you try this out and it works, please report back! (and we can add
a documentation page about it)

**Reply by mattiasmar** [reply]
I tried to catch up on the jax2tf suggestion and applied it on the [/flax/examples/imagenet/imagenet.ipynb](https://github.com/google/flax/blob/main/examples/imagenet/imagenet.ipynb) example. Code pasted below (used after cell number 16 in the notebook).
In order to get the graph from TF I need to run the model once.  I could use a little help there. Last line below fails. Any help appreciated.

```
import sonnet as snt
from jax import  random
import tree
from jax.experimental import jax2tf
def create_variable(path, value):
    name = '/'.join(map(str, path)).replace('~', '_')
    return tf.Variable(value, name=name)

class JaxModule(snt.Module):
    def __init__(self, params, apply_fn, name=None):
        super().__init__(name=name)
        self._params = tree.map_structure_with_path(create_variable, params)
        self._apply = jax2tf.convert(lambda p, x: apply_fn(p, None, x))
        self._apply = tf.autograph.experimental.do_not_convert(self._apply)

    def __call__(self, inputs):
        return self._apply(self._params, inputs)

model_cls = getattr(train.models, config.model)
model = train.create_model( model_cls=model_cls, half_precision=config.half_precision)
variables = model.init(random.PRNGKey(0), train_batch['image'])
state, params = variables.pop('params')
net = JaxModule(params, model.apply)
batch = next(iter(train_ds))
net(batch )  #This line fails
```
